x-pack/plugins/security_solution/public/common/components/threat_match/helpers.tsx

```diff
@@ -0,0 +1,162 @@
+  );
+  return {
+    field: foundField,
+    type: 'mapping',
+    value: threatFoundField,
+    entryIndex: itemIndex,
+  };
+};
+
+/**
+ * Formats the entries to be easily usable for the UI
+ *
+ * @param patterns IIndexPattern containing available fields on rule index
+ * @param entries item entries
+ */
+export const getFormattedEntries = (
+  indexPattern: IIndexPattern,
+  entries: Entry[]
+): FormattedEntry[] => {
+  return entries.reduce<FormattedEntry[]>((acc, item, index) => {
+    const newItemEntry = getFormattedEntry(indexPattern, item, index);
+    return [...acc, newItemEntry];
+  }, []);
+};
+
+/**
+ * Determines whether an entire entry or item need to be removed
+ *
+ * @param item
+ * @param entryIndex index of given entry
+ *
+ */
+export const getUpdatedEntriesOnDelete = (
+  item: ThreatMapEntries,
+  entryIndex: number
+): ThreatMapEntries => {
+  return {
+    ...item,
+    entries: [...item.entries.slice(0, entryIndex), ...item.entries.slice(entryIndex + 1)],
+  };
+};
+
+/**
+ * Determines proper entry update when user selects new field
+ *
+ * @param item - current item entry values
+ * @param newField - newly selected field
+ *
+ */
+export const getEntryOnFieldChange = (
+  item: FormattedEntry,
+  newField: IFieldType
+): { updatedEntry: Entry; index: number } => {
+  const { entryIndex } = item;
+  return {
+    updatedEntry: {
+      field: newField != null ? newField.name : '',
+      type: 'mapping',
+      value: item.value != null ? item.value.name : '',
+    },
+    index: entryIndex,
+  };
+};
+
+/**
+ * Determines proper entry update when user selects new field
+ *
+ * @param item - current item entry values
+ * @param newField - newly selected field
+ *
+ */
+export const getEntryOnThreatFieldChange = (
+  item: FormattedEntry,
+  newField: IFieldType
+): { updatedEntry: Entry; index: number } => {
+  const { entryIndex } = item;
+  return {
+    updatedEntry: {
+      field: item.field != null ? item.field.name : '',
+      type: 'mapping',
+      value: newField != null ? newField.name : '',
+    },
+    index: entryIndex,
+  };
+};
+
+export const getDefaultEmptyEntry = (): EmptyEntry => ({
+  field: '',
+  type: 'mapping',
+  value: '',
+});
+
+export const getNewItem = (): ThreatMap => {
+  return {
+    entries: [
+      {
+        field: '',
+        type: 'mapping',
+        value: '',
+      },
+    ],
+  };
+};
+
+export const filterItems = (items: ThreatMapEntries[]): ThreatMapping => {
+  return items.reduce<ThreatMapping>((acc, item) => {
+    const newItem = { ...item, entries: item.entries };
+    if (threatMap.is(newItem)) {
+      return [...acc, newItem];
+    } else {
+      return acc;
+    }
+  }, []);
+};
+
+export const containsInvalidItems = (items: ThreatMapEntries[]): boolean => {
+  return items.some((item) =>
+    item.entries.some((subEntry) => subEntry.field === '' || subEntry.value === '')
+  );
+};
+
+export const containsEmptyItem = (items: ThreatMapEntries[]): boolean => {
```

Comment:
nit: function name is a little misleading as it only checks if single-entry threat maps are empty. Passing in multiple entries will return `false` even if it contains an empty item. Totally understand what you're checking here with validation between this and `containsInvalidItems`, but might be a good idea to clear up naming/add a comment in case anyone else uses this in passing with the above assumption.

``` ts
const items: ThreatMapEntries[] = [{field: 'test', value: 'test'}, {field: '', value: '}];
containsEmptyItem(items); // returns false
```

Perhaps rename to `singleEntryThreat

Reply:
Yeah I will rename it, that sounds easy enough.